### DISCUSSION TITLE
Why this code doesn't work while it should be (or at least that's what I believe)?

### DISCUSSION THREAD

**Discussion by rempas** [MrSmith33/vox#14] [Q&A] [answered]
Hi again! I've read the spec for the attributes to try calling functions from external libraries and tied to make an example for "printf" but It seems I'm doing something wrong. I'm trying the following code:

```
enum i32 stdout = 1;
enum i32 stderr = 2;

@extern(syscall, 60)
noreturn exit(i32 error_code);

@extern(module, "libc")
i32 printf(const u8* format, ... );

i32 print(const u8[] format, ...) {
  return printf(format.ptr, )
}

void main() {
  exit(0);
}
```

When trying to compile, I'm getting the following error code: ``test.vx:7:9: Error: Expected `IDENTIFIER` after @extern(, while got `MODULE_SYM` token 'module'``

In extend, I also know that the way I'm using variadic arguments is not right but I wasn't able to find anything in the spec so can you also help with that?

P.S. The error message is being displayed in the STDOUT and not in STDERR, I don't know if you choose to do that or if you didn't noticed it.

**Comment by MrSmith33** [comment, marked as answer]
1. the whole attribute section is WIP (except `extern(syscalll)`)
2. Best place to find working examples is in test suite and existing code. See [source/tests/passing.d](https://github.com/MrSmith33/vox/blob/master/source/tests/passing.d)
3. Should all messages be in stderr?

**Reply by rempas** [reply]
> the whole attribute section is WIP (except extern(syscalll))

Oh! Does this include the attribute system itself or only the attribute section in the spec? Btw I'm seeing the addition you do to the spec and just wanted to say you are doing a great job :)

> Best place to find working examples is in test suite and existing code. See source/tests/passing.d

Oh so the work just like in D right? Got that! Thanks!

> Should all messages be in stderr?

Are you actually asking that to me to learn or hear my opinion or is this kinda of sarcastic type of question, lol? As a practice, people send their "error" (hence stderr) message to stderr so people can slip messages with error messages (so split stdout and stderr) if needed. That's not a rule but I just wanted to just point that out in case you don't want it to work like that and you didn't noticed it. However, If you want it to work like that, I don't see a problem with it. Well of course except for the fact that people cannot do what I mentioned before in the case but yeah, no real problems

**Reply by MrSmith33** [reply]
1. Only one attribute is implemented atm - `@extern(syscall)` and you can only attach it directly to the definition.
2. I guess the only time it matters where you output the text to, is when you mix different kinds of information (like data and errors) and caller wants to have only one of them.

**Reply by rempas** [reply]
> Only one attribute is implemented atm - @extern(syscall) and you can only attach it directly to the definition.

Oh, I see! Btw, this will fully support calling C functions from external libraries (.so/.dll) right? If yes then does this mean that you won't implement object files (remember the previous conversation we had)?

> I guess the only time it matters where you output the text to, is when you mix different kinds of information (like data and errors) and caller wants to have only one of them.

Yes. And well maybe when people expect it to work like that and they are doing automation, idk....

**Reply by MrSmith33** [reply]
> this will fully support calling C functions from external libraries (.so/.dll) right?

Yes, Vox uses C ABI everywhere atm, so all external functions that also use C ABI will work out of the box. Implementing object files is low priority. So you can assume no object file support for some time.

**Reply by rempas** [reply]
Nice! And imo If we already have fully support for C functions this way, they object files should not get implemented at all if they're going to reduce compilation times more than 10-15%. Of course that's an opinion and you may think differently but compilation times will be VERY import for anyone using Vox from they way I'm seeing it.

**Reply by MrSmith33** [reply]
There are only 2 usecases I want to support with .lib files: linking foreign language libraries into Vox executable/dll, or vice versa.
